src/nix-env/nix-env.cc

```diff
@@ -857,34 +859,34 @@ static VersionDiff compareVersionAgainstSet(
 
 static void queryJSON(Globals & globals, vector<DrvInfo> & elems)
 {
-    JSONObject topObj(cout, true);
+    nlohmann::json topObj;
     for (auto & i : elems) {
-        JSONObject pkgObj = topObj.object(i.attrPath);
+        nlohmann::json pkgObj = topObj[i.attrPath];
 
         auto drvName = DrvName(i.queryName());
-        pkgObj.attr("name", drvName.fullName);
-        pkgObj.attr("pname", drvName.name);
-        pkgObj.attr("version", drvName.version);
-        pkgObj.attr("system", i.querySystem());
+        pkgObj["name"] = drvName.fullName;
+        pkgObj["pname"] = drvName.name;
+        pkgObj["version"] = drvName.version;
+        pkgObj["system"] = i.querySystem();
 
-        JSONObject metaObj = pkgObj.object("meta");
+        nlohmann::json metaObj = pkgObj["meta"];
         StringSet metaNames = i.queryMetaNames();
         for (auto & j : metaNames) {
-            auto placeholder = metaObj.placeholder(j);
             Value * v = i.queryMeta(j);
             if (!v) {
                 logError({
                     .name = "Invalid meta attribute",
                     .hint = hintfmt("derivation '%s' has invalid meta attribute '%s'",
                         i.queryName(), j)
                 });
-                placeholder.write(nullptr);
+                metaObj[j] = nullptr;
             } else {
                 PathSet context;
-                printValueAsJSON(*globals.state, true, *v, placeholder, context);
+                metaObj[j] = printValueAsJSON(*globals.state, true, *v, context);
             }
         }
     }
+    std::cout << topObj;
```

Comment:
So there is a big difference between json.hpp and nlohmann/json: the former outputs JSON on the fly while the latter builds the entire JSON datastructure in memory. So the former is O(1) memory (more or less) while the latter is O(n).

Reply:
It also means that if you're piping the output into another tool, that tool won't receive any data until everything has been processed.

Reply:
Yes that was bugging me to. Some thing we do aren't too big so it doesn't really matter, but others like nar listings it might. https://github.com/nlohmann/json/discussions/2275 They are talking about streaming interfaces here, maybe I'll go pitch in.